File: 1711-find-valid-matrix-given-row-and-column-sums/find-valid-matrix-given-row-and-column-sums.py
class Solution:
    def restoreMatrix(self, rowSum: List[int], colSum: List[int]) -> List[List[int]]:
        rows = len(rowSum)
        columns = len(colSum)
        matrix = [[0] * columns for _ in range(rows)]
        
        # validate the rowsum
        for i in range(rows):
            matrix[i][0] = rowSum[i]

        # validate the columns
        for j in range(columns):
            col_sum = 0
            for i in range(rows):
                col_sum += matrix[i][j]
            
            i = 0
            while (col_sum > colSum[j]):
                difference = col_sum - colSum[j]
                max_shift = min(matrix[i][j], difference)
                matrix[i][j + 1] += max_shift
                matrix[i][j] -= max_shift
                col_sum -= max_shift
                i += 1 

        return matrix

        

File: 1711-find-valid-matrix-given-row-and-column-sums/find-valid-matrix-given-row-and-column-sums.js
var restoreMatrix = function(rowSum, colSum) {
    let rows = rowSum.length
    let columns = colSum.length 
    let matrix = Array.from({ length: rows }, () => new Array(columns).fill(0));
    let currColSum = 0

    for (let i = 0; i < rows; i++) {
        matrix[i][0] = rowSum[i]
    }  

    for (let j = 0; j < columns; j++) {
        currColSum = 0

        for (let i = 0; i < rows; i++) {
            currColSum += matrix[i][j]
        }

        let i = 0
        while (currColSum > colSum[j]) {
            let diff = currColSum - colSum[j]
            let maxShift = Math.min(diff, matrix[i][j])
            matrix[i][j + 1] += maxShift
            matrix[i][j] -= maxShift
            currColSum -= maxShift
            i += 1
        }
    }  

    return matrix


};